eip-5564.md

```diff
@@ -8,174 +8,170 @@ status: Draft
-bytes stealthPubkeyRecipient = ecAdd(pubkeyRecipient, pubkeyFromHashedSecret);
-address stealthAddressRecipient = address(stealthPubkeyRecipient);
-```
-Transfer MUST emit a `Announcement` Event containing `S` and the `stealthAddressRecipient`.
 
+/// @notice Interface for generating stealth addresses for keys on a given curve.
+/// @dev The Generator contract SHOULD have a method called decodeKeys that returns the recipient's
+/// public keys as the correct types. This is not included below since the return parameters may
+/// differ for each Generator.
+interface IERC5564Generator {
+  /// @notice Given a `registrant` and `entropy`, returns all relevant data to compute a stealth
+  /// address.
+  /// @dev MUST return all zeroes if the registrant has not registered keys for this generator.
+  /// @dev Intended to be used off-chain only to prevent exposing secrets on-chain.
+  /// @dev Consider running this against a local node, or using an off-chain library with the same
+  /// logic, instead of via an `eth_call` to a public RPC provider to avoid leaking secrets.
+  /// @dev The `entropy` is passed as `bytes memory` to support various keys formats and encryption
+  /// schemes. Consult the generator's documentation for details on how to format this parameter.
+  function generateStealthAddress(address registrant, bytes memory entropy)
+    external
+    view
+    returns (
+      address stealthAddress,
+      uint256 sharedSecret,
+      bytes32 sharedSecretHash,
+      bytes32 encryptedEntropy,
+      bytes memory ephemeralPubKey
+    );
+}
 
----
-
-
-#### For receiving (executed locally):
-
-for all `Announcement` events do:
+/// @notice Interface for sending tokens to a stealth address.
+interface IERC5564Messenger {
+  /// @dev Emitted when sending something to a stealth address.
+  /// @dev Metadata is an arbitrary field that the sender can use however they like, but the below
+  /// guidelines are recommended:
+  ///   - When sending ERC-20 tokens, the metadata SHOULD include the token address as the first 20
+  ///     bytes, and the amount being sent as the following 32 bytes.
+  ///   - When sending ERC-721 tokens, the metadata SHOULD include the token address as the first 20
+  ///     bytes, and the token ID being sent as the following 32 bytes.
+  event Announcement(
+    address indexed stealthRecipient,
+    bytes32 encryptedEntropy,
+    bytes ephemeralPubKey,
+    bytes metadata
+  );
+}
 
-if $pubtoaddr(P + (G * keccak(S * p)))$ == $stealthAddressRecipient$:
-```solidity
-bytes32 hashedSecret = keccak256(S);
-bytes sharedSecret = ecMul(hashedSecret, p);
-bytes pubkeyFromSharedSecret = ecMul(sharedSecret, G);
-bytes stealthPubkeyRecipient = ecAdd(pubkeyRecipient, pubkeyFromSharedSecret);
-address stealthAddressRecipient = address(stealthPubkeyRecipient);
-store_key_locally(p + keccak(S * p));
+/// @notice Sample implementation for generating stealth addresses for keys on the secp256k1 curve.
```

Comment:
Need to double check the math/code here is correct

Reply:
Hi Matt, I was thinking about the discussion we had about encrypting the secret vs the other method, in which we hash the shared secret and then derive the stealth address from that.
I think the latter reduces the number of parameters required.

Would look the following then:
p,P = Recipient
P*r  = Q (r = secret)
R*p = Q (R = Sender Public Key to secret)
Sender sends to: stA = Address(P + hash(Q))
Sender publishes R and stA.

Recipient does: 
R*p = Q
hash(Q)
And at this stage, we could easily implement the "view tags" method.
So instead of the recipient doing stA == Address(P + hash(p*R)), we use the one byte comparison and just do:
 hash(p*R) == stA[20:] (the remaining free bytes in the stA field can be used).
 

This is a very fundamental decission, but I don't see any downsides with the latter method and we don't need to emit the ephemeral public key.
Eventually, I'd suggest to stick with the Dual-Key Stealth Address Protocol (DKSAP), in favor of the Improved Stealth Address Protocol (ISAP), https://hackernoon.com/blockchain-privacy-enhancing-technology-series-stealth-address-i-c8a3eb4e4e43
 What do you think?

Reply:
@mds1 @apbendi I quickly implemented the above in Python. Find it [here](https://github.com/Nerolation/EIP-Stealth-Address-ERC/blob/main/minimal_poc.ipynb).

As shortly described above, I think it would be preferable to use the ISAP Stealth Addresses, allowing us to get away with one argument less.

Reply:
If we have a total of 12 bytes left for the view tag (address 20 bytes) then this should drastically reduce the parsing time (we can skip 3x EC operations) with practically no false positves.

Reply:
I'm on board with updating the scheme here to use DKSAP as spec'd out in the linked article. IIUC, the main benefit of using DKSAP over the current scheme is that it removes the need to emit the random number and therefore makes things a bit simpler. It seems to also be a bit more standardized/common than our current scheme. In terms of computation/scan time:

1. With view tags, both current scheme and DKSAP require just one EC multiplication (computing shared secret) to check if a give `Announcement` is for you, so scan performance should be similar
2. DKSAP does result in one less multiplication when computing the stealth private key, since that's now computed via addition and not multiplication. However, since that's not in the scanning critical path (it only is used when sending funds from the stealth address) this won't impact scan performance

I like the idea of using the 12 extra bytes in the address field for the view tag. 

I also found https://ieeexplore.ieee.org/document/9724375 which claims to have better scan performance than DKSAP, though I haven't read the paper yet

Reply:
This sounds nice, I agree. So we go with the ISAP scheme using 2 keys.
For the scanning part I was thinking about the following:
Since we have 12 bytes now, we can log the first 12 bytes of `$h(Q)$`, ending up reducing scanning time in `$1/(256**12)$` of the cases. 

Had a look at the paper, but I'd need more time to fully comprehend the bilinear mapping part.

I will work on changing the EIP file.

Reply:
Hi @mds1, I did some tests with viewTag and it looks like one can reduce the parsing time by 6x through it.

I wrote a quick js for that:
![image](https://user-images.githubusercontent.com/51536394/207833248-bda9a9f8-7f20-4f60-bd71-3e38f155bfa4.png)

Reply:
I plan to merge this pull request and then change some stuff discussed (ISAP with 2 keys, describe viewTag, ect)

Are you free for a quick call these days, @apbendi and @mds1?
Best!

Reply:
This is a better presentation of the simulations I carried out:
One can see that using 12 bytes of ViewTag, the parsing time can be drastically reduced.
The viewtag enables to skip 1x EC multiplication with G, 1x EC Point addition, 1x hash (keccak256). This can be done with a false positive rate of 1/(256¹²), so it can be skipped almost everytime.

The remaining operations are: 1x EC multiplication, 1x hash (keccak256). Only these operations must be carried out using ViewTags.

![image](https://user-images.githubusercontent.com/51536394/207865672-1ebb3597-5d55-49db-be48-905c70165712.png)

Reply:
Nice, 6x reduction is great! Soon I'll take the umbra scanning code and add "fake" view tag checks to simulate against the existing set of logs we currently scan against in production. Otherwise that all sounds good to me!